trinity/protocol/eth/requests.py

```diff
@@ -40,3 +48,42 @@ def validate_response(self, response: Any) -> None:
             raise ValidationError("Response must be a tuple of `BlockHeader` objects")
 
         return self.validate_headers(cast(Tuple[BlockHeader, ...], response))
+
+
+class NodeDataRequest(BaseRequest):
+    @property
+    def max_size(self) -> int:
+        return constants.MAX_STATE_FETCH
+
+    def __init__(self, node_hashes: Tuple[Hash32, ...]) -> None:
+        self.node_hashes = node_hashes
+
+    def validate_response(self, response: Any) -> None:
+        """
+        Core `Request` API used for validation.
+        """
+        if not isinstance(response, tuple):
+            raise ValidationError("Response to `HeaderRequest` must be a tuple")
+        elif not all(isinstance(item, bytes) for item in response):
+            raise ValidationError("Response must be a tuple of `BlockHeader` objects")
+        return self.validate_nodes(cast(Tuple[bytes, ...], response))
+
+    def validate_nodes(self, nodes: Tuple[bytes, ...]) -> None:
+        if not nodes:
+            # an empty response is always valid
+            return
+
+        # TODO: move to executor since keccak is expensive
+        expected_node_keys = set(self.node_hashes)
+        actual_node_keys = set(map(keccak, nodes))
+
+        if len(actual_node_keys) != len(nodes):
+            # TODO: is this something that needs to be enforced.
+            raise ValidationError("Response may not contain duplicate nodes")
+
+        unexpected_keys = actual_node_keys.difference(expected_node_keys)
+
+        if unexpected_keys:
+            raise ValidationError(
+                "Response contains {0} unexpected nodes".format(len(unexpected_keys))
```

Comment:
@gsalgado Is this correct? Should we reject a response that contains nodes that we didn't request?

Reply:
I think so, but maybe log a warning just in case a bug in our code could be causing them